# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [ ]:
def addTwo(x):
    return x + 2

print(addTwo(4))

# Lambda Function Syntax
# Example of calling a Lambda function without a variable
print((lambda x: x + 2)(4))

#### Saving to a Variable

In [1]:
add_two = lambda x: x + 2

add_two(4)

6

#### Multiple Inputs

In [2]:
# Multiple inputs with no variable
print((lambda x,y,z: x * y * z)(3,5,8))

# multiple input WITH a variable
multiple_values = lambda x,y,z: x * y * z

print(multiple_values(3,5,8))

120
120


#### Passing a Lambda into a Function

In [3]:
def multiply(f,num):
    """
        f expects to be a lambda function
        num expects a number
    """
    return f(num)

# when called, lambda will appear in first position,
# number will appear in second position

multiply(lambda x: x * x, 4)

16

#### Returning a Lambda from a Function

In [7]:
# regular defined function
def multiply_test(num):
    return num * 4

# function within a function (loosely labeled as a decorator)
def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(returnFunc())
print(f_return(4))

# Lambda function returned from a regular function
def returnLam(b,c):
    return lambda x,a: x + a + b + c

r_lamb = returnLam(4,5)
print(r_lamb(5,5))

# regular function that returns a function with variables
def return_Lamb(b,c):
    def returnedValue(x,a):
        return x + a + b + c
    return returnedValue

r_lamb = return_Lamb(4,5)
print(r_lamb(5,5))

<function returnFunc.<locals>.multiply at 0x7fe07f2a3160>
8
19
19


#### If Statements within Lambdas

In [ ]:
# lambda x: True if (condition) else False
f_condition = lambda num: num * 2 if num > 10 else num + 2

print(f_condition(8)) # adds 2

print(f_condition(12)) # multiplies by 2

print(f_condition(10)) # adds 2


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [19]:
f = lambda x: x ** 3
print(f(1), f(2), f(3))

def expon(a):
    return lambda x: a ** x
print(expon(3)(2))
print(expon(4)(3))
print(expon(2)(float('inf')))
print(expon(3)(float('-inf')))

1 8 27
9
64
inf
0.0


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [24]:
# map(func,iterable(list,dict,tuple,set))
# Normally the usage of map happens with a pre-defined function -- but we can use lambdas as well
def squared(num,num2):
    if num < 10 and num2 < 10:
        return num ** 2, num2 ** 2
    else:
        return num,num2
    
numbers = [4,11,20,3,15,20]
more_nums = [4,10,3,2,6]

squared_nums_map = list(map(squared,numbers,more_nums))
print(squared_nums_map)

lst = []
nums = [1,2,3,4,5]
def cubed(num):
    return num ** 3
    
print(list(map(cubed,nums)))


    

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]
[1, 8, 27, 64, 125]


#### Using Lambda's with Map

In [22]:
# map(Lambda x: x ** 2, iterable...)

squared_nums_lamb = list(map(lambda x, y: (x ** 2, y ** 2) if x < 10 and y < 10 else (x,y), numbers, more_nums))
print(squared_nums_lamb)

# if you have more than one iterable, you need the same number of parameters inside function

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [28]:
doubleminus_lamb = list(map(lambda x, y: (x * 2 - 1, y * 2 -1) if x > 0 and y > 0 else (x,y), numbers, more_nums))
print(doubleminus_lamb)


[(7, 7), (21, 19), (39, 5), (5, 3), (29, 11)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [31]:
# filter(function,iterable...)

names = ['Bob', 'Andy', 'Max', 'Evan', 'Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
    
new_names_list = list(filter(a_names, names))
print(new_names_list)

['Andy', 'Angelica']


#### Using Lambda's with Filter()

In [33]:
new_names_lamb = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lamb)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [48]:
nums = [2,7,4.2,1.6,9,4.4,4.9]
below_mean = list(filter(lambda num: True if num < sum(nums)/len(nums) else False, nums))
print(below_mean)

[2, 4.2, 1.6, 4.4]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [49]:
from functools import reduce

# reduce(function, iterable...)

list_1 = [2,4,6,8,10]

def addNums(num1,num2):
    return num1 + num2

result_add = reduce(addNums, list_1)
print(result_add)

30


#### Using Lambda's with Reduce()

In [50]:
result_lamb = reduce(lambda x, y: x + y, list_1)
print(result_lamb)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [53]:
my_list = [1,2,3,4]

result_lamb = reduce(lambda x, y: x * y, my_list)
print(result_lamb)

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [55]:
def addNums(nums):
    # set a base case for recursive call of function
    if nums <= 1:
        print('addNums(1) = 1')
        return nums
    else:
        # continuue to add numbers until base case is reached
        print(f'addNums({nums}) = {nums} + addNums({nums - 1})')
        return nums + addNums(nums - 1)
    
addNums(5)

addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1


15

#### Writing a Factorial Function

In [60]:
def factorial(num):
    if num == 1:
        return num
    else:
        return num * factorial(num - 1)

print(factorial(5))



120
120


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [72]:
def subnums(num):
    if num < 1:
        return num
    else:
        return num - subnums(num - 1)
print(subnums(11))

6


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [76]:
def my_range(stop,start, step = 2):
    while start < stop:
        yield start # yield keyword denotes a generator
        start += step

image_list = []
        
for i in my_range(20, 2):
    my_generator_value = i
    print(my_generator_value)
    image_list.append(my_generator_value)
    
print(image_list)

2
4
6
8
10
12
14
16
18
[2, 4, 6, 8, 10, 12, 14, 16, 18]


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [77]:
def num_sq(stop,start,step=1):
    while start >= stop:
        yield start ** 2
        start -= step
        
for i in num_sq(0, start = 3):
    print(i)

9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [84]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

not_empty = list(filter(lambda place: True if place.strip() else False, places))
print(not_empty)


['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [101]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

print(sorted(author, key=lambda name: name.lower().split()[-1]))

author.sort(key=lambda name: name.lower().split()[-1])
print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']
['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [112]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

f_places = list(map(lambda city, temp: (city, 9/5 * temp + 32), [place[0] for place in places], [place[1] for place in places]))
print(f_places)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [156]:
def printFib(num):
    
    def fib(num):
        
        if num == 0:
            return 1

        elif num == 1:
            return fib(num-1)

        else:
            return fib(num-1) + fib(num-2)
    
    for n in range(num + 1):
        print(f'Iteration {n}: {fib(n)}')
    
printFib(5)        

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
